modules/programs/waybar.nix

```diff
@@ -0,0 +1,388 @@
+          settingsModules =
+            optionalAttrs (configuration.modules != { }) configuration.modules;
+        in removeNulls (settingsWithoutModules // settingsModules);
+      # The clean list of configurations
+      finalConfiguration = map makeConfiguration cfg.settings;
+    in writePrettyJSON "waybar-config.json" finalConfiguration;
+
+    warnings = let
+      mkUnreferencedModuleWarning = name:
+        "The module '${name}' defined in '${modulesPath}' is not referenced "
+        + "in either `modules-left`, `modules-center` or `modules-right` of Waybar's options";
+      mkUndefinedModuleWarning = settings: name:
+        let
+          # Locations where the module is undefined (a combination modules-{left,center,right})
+          locations = flip filter [ "left" "center" "right" ]
+            (x: elem name settings."modules-${x}");
+          mkPath = loc: "'${modulesPath}-${loc}'";
+          # The modules-{left,center,right} configuration that includes
+          # an undefined module
+          path = concatMapStringsSep " and " mkPath locations;
+        in "The module '${name}' defined in ${path} is neither "
+        + "a default module or a custom module declared in '${modulesPath}'";
+      mkInvalidModuleNameWarning = name:
+        "The custom module '${name}' defined in '${modulesPath}' is not a valid "
+        + "module name. A custom module's name must start with 'custom/' "
+        + "like 'custom/mymodule' for instance";
+
+      # Find all modules in `modules-{left,center,right}` and `modules` not declared/referenced.
+      # cfg.settings is a list of Waybar configurations, we need to preserve the index for appropriate warnings
+      allFaultyModules = flip map cfg.settings (settings:
+        let
+          allModules = unique
+            (concatMap (x: attrByPath [ "modules-${x}" ] [ ] settings) [
+              "left"
+              "center"
+              "right"
+            ]);
+          declaredModules = attrNames settings.modules;
+          # Modules declared in `modules` but not referenced in `modules-{left,center,right}`
+          unreferencedModules = subtractLists allModules declaredModules;
+          # Modules listed in modules-{left,center,right} that are not default modules
+          nonDefaultModules = subtractLists defaultModuleNames allModules;
+          # Modules referenced in `modules-{left,center,right}` but not declared in `modules`
+          undefinedModules = subtractLists declaredModules nonDefaultModules;
+          # Check for invalid module names
+          invalidModuleNames =
+            filter (m: !isValidCustomModuleName m) (attrNames settings.modules);
+        in {
+          # The Waybar bar configuration (since config.settings is a list)
+          settings = settings;
+          undef = undefinedModules;
+          unref = unreferencedModules;
+          invalidName = invalidModuleNames;
+        });
+
+      allWarnings = flip concatMap allFaultyModules
```

Comment:
Indeed Waybar only emits warnings and accepts "invalid" configurations.
I used those variable names because I did not want to have overly long and verbose names.
I am not sure what would be the most appropriate word to convey the severity of these warnings.